### ISSUE TITLE
Tracing of tf function is stuck when the function contains custom CRF layer that wraps tfa crf functions inside

### ISSUE THREAD

**Issue by Ghostvv** [tensorflow/tensorflow#46511]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Mac OS Big Sur
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): from pypi
- TensorFlow version (use command below): 2.4
- Python version: 3.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
When wrapping `crf` functions from `tensorflow_addons` into a custom Keras layer and then creating a training function with it that will be converted into `tf.function` with dynamic first dimension of the input shape, the process is stuck during tracing when trying to create gradient tensors. if I create `tf.function` with static shapes, it works.

If I use `tfa.text.crf.crf_decode` and `tfa.text.crf.crf_log_likelihood` directly without custom Keras layer wrapper, tracing `tf.function` seems to be working.

The problem appeared in version `2.4`, it is working in `2.3`

**Describe the expected behavior**

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

```
import tensorflow as tf
import tensorflow_addons as tfa

potentials = tf.random.uniform((3,4,5))
sequence_lengths = tf.ones((3,))
tag_indices = tf.ones((3,4), dtype=tf.int32)

class CRF(tf.keras.layers.Layer):
    def build(self, input_shape: tf.TensorShape) -> None:
        self.transition_params = self.add_weight(
            shape=(5, 5),
            regularizer=tf.keras.regularizers.l2(0.1),
            name="transitions",
        )
        self.built = True
    
    def call(self, potentials, sequence_lengths):
        x = tfa.text.crf.crf_decode(
            potentials, self.transition_params, sequence_lengths
        )
        y = tfa.text.crf.crf_log_likelihood(
            potentials, tag_indices, sequence_lengths, self.transition_params
        )
        return x, y

crf = CRF()

def foo(potentials, sequence_lengths):
    with tf.GradientTape(persistent=True) as tape:
        x, y = crf(potentials, sequence_lengths)

    return x, y

tf_foo_static =  tf.function(foo)
tf_foo_static(potentials, sequence_lengths)  # works fine

tf_foo_dynamic = tf.function(foo, input_signature=[tf.TensorSpec((None, None, 5)), tf.TensorSpec((None,))])
tf_foo_dynamic(potentials, sequence_lengths)  # the execution is stuck without raising any error
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

When the process is terminated for `tfa.text.crf.crf_log_likelihood`, it looks like the gradient calculation through [`tf.scan` in `crf_forward`](https://github.com/tensorflow/addons/blob/6e7e06c56c2044e12a0d05d1381c6071defdf216/tensorflow_addons/text/crf.py#L366) is stuck, if we add `back_prop=False`, it manage to do the tracing, but it is a part of the loss function, so we need to backdrop the gradient. 

When the process is terminated for `tfa.text.crf.crf_decode`, it looks like the gradient calculation through [`tf.keras.layers.RNN` in `crf_decode_forward`](https://github.com/tensorflow/addons/blob/6e7e06c56c2044e12a0d05d1381c6071defdf216/tensorflow_addons/text/crf.py#L492) is stuck.

**Comment by Saduf2019** [comment]
@Ghostvv 
I ran the code shared on tf 2.4 and nightly,please find the [gist here](https://colab.research.google.com/gist/Saduf2019/c68946be60e2fad0eb1b06e0a173e775/untitled501.ipynb) and let us know.

**Comment by Ghostvv** [comment]
@Saduf2019 thank you very much for creating colab. Indeed `tf2.4` is stuck, while `tf2.5-nightly` seems to be working

**Comment by Saduf2019** [comment]
@Ghostvv 
Can you please upgrade and use nightly as the issue does not occur in nightly.

**Comment by Ghostvv** [comment]
I'm sorry we cannot use nightly releases. Unfortunately, as we ship `rasa` library, we have to use stable releases. Do you have a timeline when `2.5` might be released?

**Comment by ymodak** [comment]
Thanks for your issue. We are tracking this as `to-do` for upcoming TF 2.5 release. Will keep you posted as we know more.

**Comment by koernerfelicia** [comment]
@ymodak do you have any more information on this?

**Comment by ymodak** [comment]
The original code snippet is fixed with latest release TF 2.5.
See [gist](https://colab.research.google.com/gist/ymodak/3ada8cf90a99cd743f5c28cb618439d9/47450.ipynb)
